boost: Fix B2/BJam bootstrap toolchain

stmobo opened the pull request:
Boost's `bootstrap.sh` does not pass any toolchain when building B2 / BJam for its own use, which causes that part of the build to attempt to autodetect which toolchain to use.
This can cause builds to fail if the autodetected toolchain differs from the toolchain Spack wants to use.
For example, this can happen on systems with xl installed (see #20757).

This changes the install phase to first specifically build B2 / BJam using `tools/build/src/engine/build.sh`, passing an appropriate toolset parameter; the built `b2` / `bjam` executable is then passed as an additional parameter to the main `bootstrap.sh` script, and used for later parts of the install phase.

adamjstewart commented:
@hainest

hainest commented:
I don't (yet) have access to Summit, so I tested this on lassen@LLNL. gcc@8.3.1 and clang@11.0.0 work fine, but xl@16.1 fails because `build.sh` uses

    xlcpp) ( ${CXX:=xlC_r} check_cxx11.cpp

to do the C++11 compatibility check. On lassen, xlC_r uses gcc@4.9.3 for its standard library and gcc@4.9.3 requires `-std=c++11` to get a (barely) functioning c++11 compiler. I'm not sure if this same problem will show up on Summit.

@eschnett Could you try this PR there?

eschnett commented:
I tested this on Summit. I confirm that this PR corrects #20757.

ghost commented:
This patch doesn't work with boost@1.46.1 (not sure the upper/lower version for the issue).  

![image](https://user-images.githubusercontent.com/67171190/106642659-e25fa580-6545-11eb-822f-459130306766.png)


Looks like the correct path, in version 1.46.1, for line 505 is "tools/build/v2/engine/src".  Even then it will fail to find bjam.

mwkrentel commented:
@tgamblin Could we (spack) take a look at this and decide how spack
wants to handle this?  This is putting a strain on packages that use
boost, and a lot of packages use boost.

AFAIK, any system where xlC_r is on your path before your intended
compiler will cause this to break.  For example, it breaks for me on
vortex at Sandia.

IMO, clearly spack needs to pick one TOOLSET based on spack's rules
and then enforce that toolset in both phases in the boost build.

The simplest patch is the one in #20757 that restores toolset to the
way boost <= 1.74 worked.

If you look in boost issue:  https://github.com/boostorg/boost/issues/460
boost doesn't consider this a bug and thus spack will have to do
something for many future revs.  In that case, I'm inclined to use the
shortest, simplest patch, the easiest to carry forward.

But I think it's time that spack should do something and move on.

stmobo committed:
boost: Patch B2 bootstrap step to use correct toolchain

This uses the patch given in #20757, so that the B2 bootstrap
step uses the same toolset passed as an argument to bootstrap.sh.

stmobo commented:
I have a new commit, replacing the original commit in this pull request, that simply incorporates the patch in #20757 as opposed to trying to do more complicated installation step changes.

I've tested this new commit successfully on Lassen with gcc@8.4.0 and clang@11.0.0, though I couldn't test with xl@16.1 for the same reason @hainest mentioned above.

Should I push it to this PR, or should I open a new PR and close this one?

mwkrentel commented:
> Should I push it to this PR, or should I open a new PR and close this one?

Either way.  Go ahead and update this one, it will get squashed anyway.

mwkrentel commented:
@hainest and I discussed this.  We think there are two issues with
boost 1.75 and spack.  The most common and simplest is removing
`TOOLSET` as an argument from `build.sh`.  This potentially leads to
mismatched compilers, certainly with xl, but possibly with other families.

For example, on vortex at Sandia, this fails for me as:

```
2 errors found in build log:
     21    > /home/mwkrent/Vortex/spack-repo/lib/spack/env/gcc/g++ -s -O3 -qstrict -qinline -DN
           DEBUG builtins.cpp class.cpp command.cpp compile.cpp constants.cpp cwd.cpp debug.cpp
            debugger.cpp execcmd.cpp filesys.cpp frames.cpp function.cpp glob.cpp hash.cpp hcac
           he.cpp hdrmacro.cpp headers.cpp jam.cpp jamgram.cpp lists.cpp make.cpp make1.cpp md5
           .cpp mem.cpp modules.cpp native.cpp object.cpp option.cpp output.cpp parse.cpp paths
           ys.cpp regexp.cpp rules.cpp scan.cpp search.cpp jam_strings.cpp startup.cpp subst.cp
           p sysinfo.cpp timestamp.cpp variable.cpp w32_getreg.cpp modules/order.cpp modules/pa
           th.cpp modules/property-set.cpp modules/regex.cpp modules/sequence.cpp modules/set.c
           pp execunix.cpp fileunix.cpp pathunix.cpp -o b2
  >> 22    g++: error: unrecognized command line option '-qstrict'; did you mean '-Wrestrict'?
  >> 23    g++: error: unrecognized command line option '-qinline'; did you mean '-Winline'?
     24    
     25    Failed to build B2 build engine
```

Vortex is a ppc64le, xlc is on my path, not even at the front, and
`build.sh` picks the xl toolset, even though I asked for gcc.  This
results in passing xlc-style args (`-qinline`) to gcc/g++.

I tested vortex, and the patch in this PR restores the behavior from
<= 1.74 and fixes the problem.

```
-  (cd "$my_dir/tools/build/src/engine" && ./build.sh)
+  (cd "$my_dir/tools/build/src/engine" && ./build.sh "$TOOLSET")
```

@hainest you were going to check lassen at LLNL.

We could ask @eschnett to retest summit, but this patch is now the
same as his from #20757, it worked before, so it should work again.

----------

The second problem (which I haven't seen) is what @hainest reported above.

```
xlcpp) ( ${CXX:=xlC_r} check_cxx11.cpp
```

----------

So, I'm going to suggest that we commit this, the simple fix to the
more common problem now.  Then, @hainest, if you like, can propose
an additional fix to the second problem as needed.

OK ?

hainest commented:
This revised patch (4ced9a7061b55640478317360301212b1ad1c3d9) is working on lassen as [before](https://github.com/spack/spack/pull/21408#issuecomment-770296187).

@adamjstewart Given @mwkrentel's results, I think we are ready to merge.

mwkrentel commented:
Thanks!  This makes my life simpler for my project (hpctoolkit) where I don't
have to tip-toe around bad revs of dependency packages.
